pkg/controllers/securitypolicy_controller.go

```diff
@@ -27,82 +29,108 @@ import (
+}
+
+func deleteFail(r *SecurityPolicyReconciler, c *context.Context, o *v1alpha1.SecurityPolicy, e *error) {
+	r.setSecurityPolicyReadyStatusFalse(c, o, e)
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteFailTotal, METRIC_RES_TYPE)
+}
+
+func updateSuccess(r *SecurityPolicyReconciler, c *context.Context, o *v1alpha1.SecurityPolicy) {
+	r.setSecurityPolicyReadyStatusTrue(c, o)
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateSuccessTotal, METRIC_RES_TYPE)
+}
+
+func deleteSuccess(r *SecurityPolicyReconciler, c *context.Context, o *v1alpha1.SecurityPolicy) {
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteSuccessTotal, METRIC_RES_TYPE)
+}
+
 func (r *SecurityPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	obj := &v1alpha1.SecurityPolicy{}
 	log.Info("reconciling securitypolicy CR", "securitypolicy", req.NamespacedName)
 	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerSyncTotal, METRIC_RES_TYPE)
 
 	if err := r.Client.Get(ctx, req.NamespacedName, obj); err != nil {
-		log.Error(err, "unable to fetch security policy CR", "req", req.NamespacedName)
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+		log.V(1).Info("fetch security-policy CR", "req", req.NamespacedName)
+		return resultNormal, client.IgnoreNotFound(err)
 	}
 
 	if obj.ObjectMeta.DeletionTimestamp.IsZero() {
 		metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateTotal, METRIC_RES_TYPE)
 		if !controllerutil.ContainsFinalizer(obj, util.FinalizerName) {
 			controllerutil.AddFinalizer(obj, util.FinalizerName)
 			if err := r.Client.Update(ctx, obj); err != nil {
-				metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateFailTotal, METRIC_RES_TYPE)
-				return ctrl.Result{}, err
+				log.Error(err, "add finalizer", "security-policy", req.NamespacedName)
+				updateFail(r, &ctx, obj, &err)
+				return resultRequeue, nil
 			}
 			log.V(1).Info("added finalizer on securitypolicy CR", "securitypolicy", req.NamespacedName)
 		}
 
 		if isCRInSysNs, err := r.isCRRequestedInSystemNamespace(&ctx, &req); err != nil {
-			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateFailTotal, METRIC_RES_TYPE)
-			return ctrl.Result{}, err
+			err = errors.New("fetch namespace associated with security policy CR failed")
+			log.Error(err, "would retry exponentially", "security-policy", req.NamespacedName)
+			updateFail(r, &ctx, obj, &err)
+			return resultRequeue, nil
 		} else if isCRInSysNs {
-			err = errors.New("Security Policy CR cannot be created in System Namespace")
-			log.Error(err, "failed to create security policy CR", "securitypolicy", req.NamespacedName)
-			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateFailTotal, METRIC_RES_TYPE)
-			return ctrl.Result{}, err
+			err = errors.New("security Policy CR cannot be created in System Namespace")
+			log.Error(err, "", "security-policy", req.NamespacedName)
```

Comment:
{"security-policy": "test", "error": "security Policy CR cannot be created in System Namespace"}
I think this info is enough? Since err already specify the message.